### ISSUE TITLE
Training on top of MsCOCO+ Custom class

### ISSUE THREAD

**Issue by abhigoku10** [dbolya/yolact#64]
@dbolya @fanyix @chongzhou96 @zllrunning  Thanks for the wonderful code . I am having few queries 
1.How much time did it too to train the network for Mscoco dataset from scratch
2. What are the types of augmentation you have applied 
3. What is final number of images used as a training set
4. Can i freeze the layers and add additional classes to MsCOCO(from 80 to 90) and re-train it, if so whats is the procedure to do it

**Comment by dbolya** [comment]
I'm going to need more info than that, bruv.

**Comment by dbolya** [comment]
1. 4 Days (on 1 P100, should take ~5 on something like a Titan Xp)
2. Copying from `data/config.py`:
```python
    # SSD data augmentation parameters
    # Randomize hue, vibrance, etc.
    'augment_photometric_distort': True,
    # Have a chance to scale down the image and pad (to emulate smaller detections)
    'augment_expand': True,
    # Potentialy sample a random crop from the image and put it in a random place
    'augment_random_sample_crop': True,
    # Mirror the image with a probability of 1/2
    'augment_random_mirror': True,
```
See `utils/augmentations.py` for implementation details.
3. ~118k (COCO's 2017 train split). I can fetch the exact number if you want.
4. Yup, and it's pretty easy too. Just follow the custom dataset procedure in the ReadMe and follow #36 to fine tune from the existing weights.

**Comment by abhigoku10** [comment]
thanks for the quick response , i have a query on point 5 for which i shall elaborate .
The yolact_base_54_800000.pth was trained on MS coco for 80 classes , i want to perform custom training which is of 20 classes along with 80 classes of MSCOCo , since training from scratch takes time i want to freeze the max conv layers so that the network will not forget the previous information of 80 classes along with it learn the next 20 classes

**Comment by dbolya** [comment]
You don't need to freeze the previous weights, just load them in place of the pretrained imagenet weights. So essentially you start training from where I left off training when I uploaded those weights. That's what #36 does. Also now that you're fine-tuning, you don't need to train for a full 800k iterations so you can change `max_iter` to like 200k then change `lr` to start lower (let's say `1e-4`). Thus it'll take 1/4 the time (these parameters aren't optimal so it might converge before then).

**Comment by abhigoku10** [comment]
Thanks for the response shall try and respond back to you

**Comment by abhigoku10** [comment]
Hi, I tried training the model as suggested, the training is executed without any errors but when I test the model it has learnt only the new 10 class but has forgotten or not predicting the other 80 classes from MSCoco

**Comment by dbolya** [comment]
Did you add your new data to COCO and train with the combined training set, or did you just train with your new data. If the latter, then it's going to forget everything it learned about the old data (and there's no easy way to freeze some part so it doesn't forget atm so you have to train with both).

**Comment by abhigoku10** [comment]
Thanks for response, I did the later part so I am correcting it and going to re-train again with both mscoco data and my new data

**Comment by abhigoku10** [comment]
@dbolya thanks for the response now i am getting the expected results , but have few additional queries
1.Since the MScoco have huge data , for the model to learn the new custom data . what is the minimum number of dataset required 
2. will there be any change in the performance of the base model when new data is added and re-trained , if so what should be done to obtain good accuracy 
Thanks in advance

**Comment by dbolya** [comment]
1. Not sure. Perhaps around 100-1k examples per class? Maybe you can just sample 1/10th of the images randomly, or bring the number of examples in each of the original classes in line with your new classes.

2. Also not sure, since I haven't tested training on new classes. I'd expect a small drop in performance for the old classes. However, if the new classes are not that similar to the old ones, adding them probably wouldn't have much effect.

**Comment by annael99** [comment]
@abhigoku10 could i see the code where you train on the 90/100 classes? Also, i'd love to hear about the results you got (performance-wise and speed-wise) if it possible.
Thanks in advance

**Comment by abhigoku10** [comment]
@annael99 hi i would not be able to share the code due to company rules, but i can other information . The speed wise it was similar as mentioned in the paper and by the accuracy wise it had droped 2-3% compared ot the orginal values

**Comment by annael99** [comment]
@abhigoku10 thanks a lot, even the custom dataset cannot be shared? I would love to see the output by myself haha

**Comment by Gowtham2305** [comment]
@abhigoku10 
1) I used this reference to create a dataset **https://github.com/dbolya/yolact/issues/70#issuecomment-504283008**
2) I have to train the custom dataset (8 new classes) + 80 Classes from MSCOCO so that network should able to detect 88 classes in total.
3) Is it possible to provide a reference on how you generated a combined training dataset?(MSCOCO + Custom dataset).

**Comment by abhigoku10** [comment]
@Gowtham2305  to train  custom dataset (8 new classes) + 80 Classes from MSCOCO u need to add MSCOCO images along with new images and add the additional class with ID in the training , sorry i do not have any formal doc

**Comment by Gowtham2305** [comment]
@abhigoku10 thank you for the response. when I created a custom dataset using https://github.com/dbolya/yolact/issues/70#issuecomment-504283008  I will get  a annotations.json and JPEGImages . How will I use this annotations.json along with the COCO's  instances_train2017.json during training?If you could provide steps it will be helpful.
